checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java

```diff
@@ -0,0 +1,192 @@
+package org.checkerframework.checker.mustcall;
+
+import com.sun.source.tree.AnnotationTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.ReturnTree;
+import com.sun.source.tree.Tree;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import org.checkerframework.checker.compilermsgs.qual.CompilerMessageKey;
+import org.checkerframework.checker.mustcall.qual.MustCallAlias;
+import org.checkerframework.checker.mustcall.qual.NotOwning;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.basetype.BaseTypeVisitor;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedDeclaredType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType;
+import org.checkerframework.javacutil.ElementUtils;
+import org.checkerframework.javacutil.TreePathUtil;
+import org.checkerframework.javacutil.TreeUtils;
+
+/**
+ * The visitor for the Must Call checker. This visitor is similar to BaseTypeVisitor, but overrides
+ * methods that don't work well with the MustCall type hierarchy because it doesn't use the top type
+ * as the default type.
+ */
+public class MustCallVisitor extends BaseTypeVisitor<MustCallAnnotatedTypeFactory> {
+
+  /**
+   * Creates a new MustCallVisitor.
+   *
+   * @param checker the type-checker associated with this visitor
+   */
+  public MustCallVisitor(BaseTypeChecker checker) {
+    super(checker);
+  }
+
+  @Override
+  public Void visitReturn(ReturnTree node, Void p) {
+    // Only check return types if ownership is being transferred.
+    MethodTree enclosingMethod = TreePathUtil.enclosingMethod(this.getCurrentPath());
+    // enclosingMethod is null if this return site is inside a lambda. TODO: handle lambdas more
+    // precisely?
+    if (!checker.hasOption(MustCallChecker.NO_LIGHTWEIGHT_OWNERSHIP) && enclosingMethod != null) {
```

Comment:
Put the option test before the call to `enclosingMethod`, which is not cheap.
Also, I suggest having a boolean field in this, so that `hasOption` can be called just once at construction time.

Reply:
Can you respond to [my comment](https://github.com/typetools/checker-framework/pull/4484#discussion_r610949660) about `hasOption` from the previous round of reviews?